### ISSUE TITLE
ArrayTypeMismatchException when decoding Opus frames 40 bytes or less

### ISSUE THREAD

**Issue by bentorkington** [lostromb/concentus#34]
using v1.1.7 from NuGet, and a variety of Axis manufactured cameras and audio bridges, I'm experiencing a crash whenever using bitrates 16kb/s and lower. This seems to reproduce on any frame, and is resolved by choosing a bitrate 24kb/s or higher.

in: Concentus.Silk.DecodeAPI.silk_Decode(): ArrayTypeMismatchException: Source array type cannot be assigned to destination array type.

Checking the sender by streaming from it with VLC suggests that valid Opus frames are being sent at these low bitrates. Please let me know if you'd like some sample frames / anything else I can do to help.

**Comment by bentorkington** [comment]
Perhaps I could capture a bunch of packets into a raw dump for adding to the test suite?

**Comment by aenz-bob** [comment]
Free free, it is the roadside camera and mic.

**Comment by lostromb** [comment]
You can start by listing all the parameters you used in encoding and maybe some code snippets.

This seems strange because the parity tester examined thousands of combinations of bitrates and modes and never saw an exception like this. But maybe there's something about the codec setup that's unique here.

**Comment by bentorkington** [comment]
We don't directly control the encoding except through the web interface of the Axis camera / audio bridge. There, we can choose the sample rate and stream bit rate. Irrespective of sample rate, the following code works whenever a bit rate of 24kbit/s or more is used, and always fails whenever a bit rate of 16kbit/s or below is used.

We're using the SharpRTSP library to stream Opus audio from the RTSP source. `Client_Received_Opus` is fired once with each RTP frame's data.

`WaveBuffer` is a union-like struct from the NAudio project that helps convert between arrays of samples as `short`s and `byte`s. https://markheath.net/post/wavebuffer-casting-byte-arrays-to-float

```csharp
public Opus(RTSPClient client, int sampleRate, int nChannels, SourceLogger logger) : base(logger)
{
    // RFC7587 7. SDP Considerations:
    //
    // The media subtype ("opus") goes in SDP "a=rtpmap" as the encoding
    // name.The RTP clock rate in "a=rtpmap" MUST be 48000, and the
    // number of channels MUST be 2.

    if (sampleRate != 48000 || nChannels != 2)
        throw new ArgumentException("invalid RTP payload for Opus");

    // Force stereo output from the decoder for now, as our WMA encoder insists on Stereo
    decoder = OpusDecoder.Create(sampleRate, nChannels);
    decodeBuffer = new byte[960 * 2 /* bytes per sample */ * 2 /* channels */];
    waveBuffer = new WaveBuffer(decodeBuffer);
    client.Received_Opus += Client_Received_Opus;
}

private void Client_Received_Opus(string format, byte[] opus, uint timestamp)
{
    // Here, byte[] opus is 40 bytes long for 16kbit/s streams, and 60 bytes long for 24kbit/s streams.
    // Whenever byte[] opus is 40 bytes or less, ArrayTypeMismatchException is thrown inside silk_Decode()

    int decodedSamples = decoder.Decode(opus, 0, opus.Length, waveBuffer.ShortBuffer, 0, 960);
    /* (process the PCM data) */
}
```

**Comment by lostromb** [comment]
Okay, I get it now. It's triggering at low bitrates because that hits a specific code path in the SILK codec that uses `Array.Copy` to move data, and that throws an exception when the underlying array types are unexpected. I remember I've dealt with this before when working with NAudio and the workaround was to use `Buffer.BlockCopy` internally instead, which doesn't care about the underlying array types as long as they were blittable. Strange though that I can't reproduce it locally...

The simple workaround would be to decode the packet into a separate strongly-typed `short[]` array that you allocate, and then `Buffer.BlockCopy(tempBuffer, 0, waveBuffer.ShortBuffer, 0, decodedSamples * 2 /* numChannels */ * sizeof(short))` afterwards to get it into the NAudio buffer. But I can take this as feedback for the `modernize` branch that I'm working in which will likely get rid of all the internal `Array.Copy` calls anyways because they're not the fastest method available.

**Comment by bentorkington** [comment]
Thanks Logan, can confirm this fixes it. We really appreciate your help!

I've made a minimally reproducible example that uses NAudio's WaveBuffer short <=> byte magic. I was wondering why I couldn't reproduce it with OpusEncoder at first, and then your message came through and it all became clear.

My repro is at https://github.com/bentorkington/ConcentusLowBitrate if it helps.

Once again, thanks.

**Comment by lostromb** [comment]
Should be fixed for good in the 2.0.0 release, as I have removed all of the offending internal calls to `Array.Copy`
